File: generators/lib/build_environment.py
"""Build-environment provenance capture (issue #41).

Complements the existing generator-script-hash provenance (see
`release_yaml.py`'s module docstring) with the software environment that
actually executed a generator: this repository's commit, the immutable
`opengwasdb` revision installed, the `pixi.lock` this environment was
resolved from, which Pixi environment ran the command, and the interpreter
version. Best-effort throughout — a field that can't be determined (e.g. not
running under Pixi, or `opengwasdb` not installed via its pinned Git commit)
is recorded as `unavailable` rather than raising, since this is provenance
metadata, not something a build should fail over.
"""
from __future__ import annotations

import json
import platform
import subprocess
from pathlib import Path

UNAVAILABLE = "unavailable"


def _run(args: list[str]) -> str | None:
    try:
        result = subprocess.run(args, capture_output=True, text=True, timeout=10)
    except (OSError, subprocess.TimeoutExpired):
        return None
    if result.returncode != 0:
        return None
    return result.stdout.strip()


def _repo_commit(root: Path) -> str:
    return _run(["git", "-C", str(root), "rev-parse", "HEAD"]) or UNAVAILABLE


def _pixi_lock_sha256(root: Path) -> str:
    lock = root / "pixi.lock"
    if not lock.exists():
        return UNAVAILABLE
    import hashlib

    digest = hashlib.sha256()
    digest.update(lock.read_bytes())
    return digest.hexdigest()


def _opengwasdb_commit() -> str:
    """Best-effort: pip records the resolved Git commit of a `git+https://...`
    dependency in the installed distribution's `direct_url.json`."""
    try:
        import importlib.metadata as metadata

        dist = metadata.distribution("opengwasdb")
        direct_url_text = dist.read_text("direct_url.json")
        if not direct_url_text:
            return UNAVAILABLE
        direct_url = json.loads(direct_url_text)
        commit = direct_url.get("vcs_info", {}).get("commit_id")
        return commit or UNAVAILABLE
    except Exception:
        return UNAVAILABLE


def _opengwasdb_version() -> str:
    try:
        import importlib.metadata as metadata

        return metadata.version("opengwasdb")
    except Exception:
        return UNAVAILABLE


def capture(root: Path) -> dict[str, str]:
    """Capture this process's build-environment provenance, rooted at `root`
    (the opengwasdb-stores repository root)."""
    import os

    return {
        "repo_commit": _repo_commit(root),
        "opengwasdb_version": _opengwasdb_version(),
        "opengwasdb_commit": _opengwasdb_commit(),
        "pixi_lock_sha256": _pixi_lock_sha256(root),
        "pixi_environment": os.environ.get("PIXI_ENVIRONMENT_NAME", UNAVAILABLE),
        "platform": platform.platform(),
        "python_version": platform.python_version(),
    }


if __name__ == "__main__":
    print(json.dumps(capture(Path(__file__).resolve().parents[2]), indent=2, sort_keys=True))

File: resources/generators/lib/build_environment.R
# Build-environment provenance capture (issue #41), R side.
#
# Mirrors resources/generators/lib/build_environment.py: complements the existing
# generator-script-hash provenance with the software environment that
# actually ran the generator (repo commit, pixi.lock hash, Pixi environment
# name, tool versions). Best-effort — any field that can't be determined is
# recorded as "unavailable" rather than aborting the build.

UNAVAILABLE <- "unavailable"

.bee_run <- function(args) {
  result <- tryCatch(
    system2(args[1], args[-1], stdout = TRUE, stderr = TRUE),
    error = function(e) NULL,
    warning = function(w) NULL
  )
  status <- attr(result, "status")
  if (is.null(result) || (!is.null(status) && status != 0)) return(NULL)
  result
}

.bee_repo_commit <- function(root) {
  out <- .bee_run(c("git", "-C", root, "rev-parse", "HEAD"))
  if (is.null(out)) UNAVAILABLE else out[1]
}

.bee_pixi_lock_sha256 <- function(root) {
  lock <- file.path(root, "pixi.lock")
  if (!file.exists(lock)) return(UNAVAILABLE)
  out <- .bee_run(c("sha256sum", lock))
  if (is.null(out)) UNAVAILABLE else strsplit(out[1], "\\s+")[[1]][1]
}

.bee_tool_version <- function(cmd, flag = "--version") {
  path <- Sys.which(cmd)
  if (identical(unname(path), "")) return(UNAVAILABLE)
  out <- .bee_run(c(cmd, flag))
  if (is.null(out) || length(out) == 0) UNAVAILABLE else out[1]
}

#' Capture this process's build-environment provenance, rooted at `root`
#' (the opengwasdb-stores repository root).
build_environment_info <- function(root) {
  list(
    repo_commit = .bee_repo_commit(root),
    pixi_lock_sha256 = .bee_pixi_lock_sha256(root),
    pixi_environment = { v <- Sys.getenv("PIXI_ENVIRONMENT_NAME", unset = NA); if (is.na(v) || v == "") UNAVAILABLE else v },
    platform = R.version$platform,
    r_version = R.version.string
  )
}
